### ISSUE TITLE
The ComboBox doesn't select value if it's enum

### ISSUE THREAD

**Issue by alexdi220** [microsoft/microsoft-ui-xaml#5922]
**Describe the bug**
Hi, the binding to the ComboBox with `SelectedValuePath` doesn't work when I've placed the enum to `SelectedValue`. Please take a look at the following code.

**Steps to reproduce the bug**
```xaml
    <StackPanel Orientation="Horizontal" HorizontalAlignment="Center" VerticalAlignment="Center">
        <ComboBox SelectedValuePath="Value" 
                  DisplayMemberPath="Text"
                  ItemsSource="{x:Bind TestSource}"
                  SelectedValue="{x:Bind MyValue, Mode=TwoWay}"/>
    </StackPanel>
```
```csharp
    public sealed partial class MainWindow : Window {
        public MainWindow() {
            TestSource = new DisplayItem[] {
                new DisplayItem(MyValues.Value1, "Value1"),
                new DisplayItem(MyValues.Value2, "Value2"),
            };
            MyValue = TestSource[0].Value;

            this.InitializeComponent();
        }

        public IList<DisplayItem> TestSource { get; }
        public MyValues MyValue { get; set; }
    }

    public enum MyValues { Value1, Value2 }

    public class DisplayItem {
        public DisplayItem(MyValues value, string text) {
            Value = value;
            Text = text;
        }

        public MyValues Value { get; }
        public string Text { get; }
    }
```

**Expected behavior**

**Screenshots**
![image](https://user-images.githubusercontent.com/47348109/133961185-af4bab60-4b03-488c-b424-58201e1d02b0.png)

![image](https://user-images.githubusercontent.com/47348109/133961137-c87bc1fc-a552-477b-8069-2316bcd950e2.png)

**Version Info**

NuGet package version: 

[WinUI 3 - Windows App SDK 0.8: 0.8.0] If you're seeing your issue in older previews of WinUI 3, please try this release.

Windows app type:
| UWP              | Win32            |
| :--------------- | :--------------- |
|  | Yes |

| Windows version                    | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)              |    |
| May 2021 Update (19043)            |    |
| October 2020 Update (19042)        |    |
| May 2020 Update (19041)            |  Yes   |
| November 2019 Update (18363)       |    |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            | Yes |
| Xbox               |  |
| Surface Hub        |  |
| IoT                |  |

**Additional context**

**Comment by GeorgeKerwood** [comment]
I may be wrong but I think this is due to your usage of the ComboBox. You've assigned the `ItemsSource` as a collection of `DisplayItem` and you're expecting the `SelectedItem` to be of type `MyValues`. Simply, `MyValue` isn't a member of `TestSource` so how could it be the selected item?

**Comment by alexdi220** [comment]
> without setting the SelectedValuePath
I've set it `<ComboBox SelectedValuePath="Value" `

**Comment by movpushmov** [comment]
Same issue

**Comment by GeorgeKerwood** [comment]
I'm not convinced this is enum specific or even a bug, more an issue of the binding approach. 

Consider that you've bound `SelectedValue` as `TwoWay`: What would your expected behaviour be on a code-behind change to `MyValue` be? Would the ComboBox selected the first item whereby it's `Value` property was equal to `MyValue`? Or the last? What if no items in `TestSource` had a `Value` property was equal to `MyValue`? Enum, string or any other type this will not work, not to mention that you haven't implemented `INotifyPropertyChanged`. I could see this working if the binding mode ["OneWayToSource"](https://docs.microsoft.com/en-us/dotnet/api/system.windows.data.bindingmode?view=net-5.0) was available but curiously it is not. 

Consider the following implementation instead which works just fine:
```xaml
<StackPanel Orientation="Horizontal" HorizontalAlignment="Center" VerticalAlignment="Center">
    <ComboBox DisplayMemberPath="Text"
              ItemsSource="{x:Bind TestSource}"
              SelectedItem="{x:Bind SelectedItem, Mode=TwoWay}"/>
</StackPanel>
```
```c#
public sealed partial class MainWindow : Window, INotifyPropertyChanged
{
    public MainWindow()
    {
        TestSource = new DisplayItem[] {
            new DisplayItem(MyValues.Value1, "Value1"),
            new DisplayItem(MyValues.Value2, "Value2"),
        };
        SelectedItem = TestSource[0];

        this.InitializeComponent();
    }

    public IList<DisplayItem> TestSource { get; }
    
    private DisplayItem selectedItem;
    public DisplayItem SelectedItem
    {
        get => selectedItem;
        set
        {
            if (value != selectedItem) { selectedItem = value; NotifyPropertyChanged("SelectedItem"); }
        }
    }

    public MyValues MyValue { get => SelectedItem.Value; }

    public event PropertyChangedEventHandler PropertyChanged;
    private void NotifyPropertyChanged(string propertyName)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }
}
```

I have a feeling that perhaps your example here is an over simplification and maybe you were trying to do something specific and for a good reason that I've missed? However, the little xaml snippet below will show that the ComboBox **does** indeed select value if it's enum:
```xaml
<StackPanel Orientation="Horizontal" HorizontalAlignment="Center" VerticalAlignment="Center">
    <ComboBox Name="myComboBox"
              SelectedValuePath="Value" 
              DisplayMemberPath="Text"
              ItemsSource="{x:Bind TestSource}"/>
    <TextBox Text="{Binding ElementName=myComboBox, Path=SelectedValue}"/>
</StackPanel>
```

**Comment by StephenLPeters** [comment]
Also try InitializeCompenent call before doing other things.

**Comment by alexdi220** [comment]
@GeorgeKerwood Did you try my code? It's doesn't work. I know many ways to bind the SelectedItem. And your approach will work. But it has a lot of excessive code. The `SelectedValuePath` property exists only for one thing - to select a value instead of a whole item. This is a very simple and easy scenario when the ComboBox does "work" instead of you. In this case, you no need to write a code that syncs a real value in the viewmodel\model and the SelectedItem in the ComboBox. It's a basic feature.

**Comment by alexdi220** [comment]
> Also try InitializeCompenent call before doing other things.

I guess it makes my sample work but this is a workaround. I've already rewritten my code (in a day when I've posted the issue). All I need is a ComboBox to select enum value but the Items should be localized.

**Comment by GeorgeKerwood** [comment]
Hi @alexdi220,

Yes, I did try your code, and no, it's not working. I've also tried the same in a "traditional" WPF application and yes, it does work, **BUT** it automatically treats the binding mode as "OneWayToSource", meaning a selection change on the ComboBox will update your `MyValue` property but a `MyValue` change in code-behind will not change the ComboBox selection and nor should it as, consider the scenario:
```c#
TestSource = new DisplayItem[] {
    new DisplayItem(MyValues.Value1, "Value1"),
    new DisplayItem(MyValues.Value2, "Value2"),
    new DisplayItem(MyValues.Value2, "Value3"),
    new DisplayItem(MyValues.Value1, "Value4"),
};

MyValue = MyValues.Value2;
```
How would the control know to select index 1 or index 2 since they both have a `Value` of `MyValues.Value2`? A `TwoWay` binding cannot work on `SelectedValue` whilst targeting a sub-property of the item since it's often ambiguous. 

Now... for some reason or other, it would seem that WinUI 3 does not have a "OneWayToSource" binding mode to default to, only "OneWay", "OneTime" and "TwoWay". Since none of these would resolve the ambiguity issue, I believe this is why the binding is not being applied.

You'll find the shame behaviour regardless of type, this is not enum specific. You'll also see if you try the xaml snippet I included at the end of my first response; the ComboBox **IS** selecting the value, it's just the binding that's not working.

Maybe you've stumbled on a bigger issue here that "OneWayToSource" binding is not supported?

**Comment by alexdi220** [comment]
The `enum` is a value type. You can try to change the `public MyValues MyValue { get; set; }` to `public int MyValue { get; set; }` (DisplayItem class needs a change too). It will work. 
As for your last sample:
- User selects item "Value2" -> SelectedValue=MyValues.Value2
- User selects item "Value3" -> SelectedValue=MyValues.Value3
- The viewmodel changes property Value to MyValues.Value2 ->ComboBox.SelectedValue=MyValues.Value2 (first which can be found)

[SelectedValuePath in WPF](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.primitives.selector.selectedvaluepath?redirectedfrom=MSDN&view=net-5.0#System_Windows_Controls_Primitives_Selector_SelectedValuePath)

**Comment by GeorgeKerwood** [comment]
Yes, you're right, a change to `int` means the ComboBox initialises with a selected item according to the value of `MyValue`. However, code-behind changes to `MyValue` still do not result in ComboBox selection changes.

**Comment by alexdi220** [comment]
Yes, it's expected. The MainWindow doesn't implement the INotifyPropertyChanged. The code is just a sample.

**Comment by GeorgeKerwood** [comment]
Then I apologize for more of an interference than a help. I understood this as your problem, not the fact that the ComboBox’s initial selection is failing. My mistake.

**Comment by DRAirey1** [comment]
This is a legitimate bug. As one person noted, **enum** is a value type. It has an 'IsEqual' and 'Hash' method, so there's no reason an item can't be found in an *ItemsSource* that provides the same type of value.

**Comment by mikernet** [comment]
This should not be closed.

**Comment by mikernet** [comment]
@bpulliam Can you reopen this or do I need to create a new issue?

**Comment by kmahone** [comment]
Dupe of #8919

**Comment by DRAirey1** [comment]
> Dupe of #8919

How can you say this when this issue predates that one by two years? #8919 is a dupe of this issue.
